libs/template/materialize.go

```diff
@@ -56,23 +56,23 @@ func Materialize(ctx context.Context, configFilePath, templateRoot, outputDir st
 		}
 	}
 
+	r, err := newRenderer(ctx, config.values, helpers, templatePath, libraryPath, outputDir)
+	if err != nil {
+		return err
+	}
+
 	// Prompt user for any missing config values. Assign default values if
 	// terminal is not TTY
-	err = config.promptOrAssignDefaultValues()
+	err = config.promptOrAssignDefaultValues(r)
 	if err != nil {
 		return err
 	}
-
 	err = config.validate()
 	if err != nil {
 		return err
 	}
 
 	// Walk and render the template, since input configuration is complete
-	r, err := newRenderer(ctx, config.values, helpers, templatePath, libraryPath, outputDir)
```

Comment:
I love the idea here but I am not sure we should be doing it this way from an implementation POV.

My concerns here:
Right now we have this invariant that the `renderer` holds a validated config. With this approach this no longer holds true. This is fine but I think we should be more deliberate by:
1.  validating the partial config we use when rendering descriptions / default values For example we should error if someone passes an string value for a property that expects an integer during description / default value computation.
2. Error if a description / default value uses a variable that does not have a value yet. We probably are already doing this.

Separately, we should also consolidate the config and renderer structs into one, which we can do by including the `schema` in the renderer struct. Could possibly be a followup.

Reply:
Thanks @shreyas-goenka! I updated this PR to do the extra validations.